torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1884,6 +1884,41 @@ def _lazy_init(self) -> None:
         # to non-root instances
         inconsistent_limit_all_gathers = False
         for fsdp_module in self.fsdp_modules(self):
+            if not fsdp_module._use_orig_params and fsdp_module._has_params:
+                # Check if the wrapped module changed after construction
+                # (e.g. applying the activation checkpointing wrapper) and
+                # if so, de-register the `FlatParameter` from the old
+                # wrapped module and register it to the new wrapped module
+                flat_param = fsdp_module._handles[0].flat_param
+                target_submodule = None
+                target_name = None
+                for submodule in fsdp_module.modules():
+                    for param_name, param in submodule._parameters.items():
+                        if flat_param is param:
+                            target_submodule = submodule
+                            target_name = param_name
+                            break
+                if (
+                    target_submodule is not None
+                    and target_submodule is not fsdp_module.module
+                ):
+                    assert target_name is not None
+                    if fsdp_module._debug_level == dist.DebugLevel.DETAIL:
+                        warnings.warn(
+                            "The FSDP wrapped module changed from "
+                            f"{target_submodule} to {fsdp_module.module} on "
+                            f"rank {fsdp_module.rank}. {fsdp_module}"
+                        )
+                    target_submodule._parameters.pop(target_name)  # de-register
+                elif target_submodule is None:
+                    raise RuntimeError(
+                        "Either the FSDP wrapped module was removed from "
+                        "the model or its `FlatParameter` was manually "
+                        f"de-registered on rank {fsdp_module.rank}. Both of "
+                        f"these are invalid behavior. {fsdp_module}"
+                    )
+                if target_submodule is not fsdp_module.module:
+                    fsdp_module._register_flat_param()
```

Comment:
Can this line be moved after line 1912, the if block in line 1901?

Reply:
Yes, good suggestion! I originally did not raise `RuntimeError` in the case that `target_submodule is None` and tried to still re-register, which was why this re-registration was here below. However, I realized that something would likely break soon after, so I chose to raise `RuntimeError`.